Source/WTF/wtf/text/WYHasher.h

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (C) 2005-2017 Apple Inc. All rights reserved.
+ * Copyright (C) 2010 Patrick Gansterer <paroga@paroga.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ *
+ */
+
+#pragma once
+
+#include <cstdio>
+#include <stdint.h>
+#include <string.h>
+#include <unicode/utypes.h>
+#include <wtf/Assertions.h>
+#include <wtf/FastMalloc.h>
+#include <wtf/Vector.h>
+#include <wtf/text/LChar.h>
+
+namespace WTF {
+
+class WYHasher {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static constexpr const unsigned flagCount = 8;
+    static constexpr const unsigned maskHash = (1U << (sizeof(unsigned) * 8 - flagCount)) - 1;
+    static constexpr const uint64_t secret[4] = { 0xa0761d6478bd642full, 0xe7037ed1a0b428dbull, 0x8ebc6af09c88c6e3ull, 0x589965cc75374cc3ull };
+
+    WYHasher() = default;
+
+    template<typename T, typename Converter>
+    static unsigned computeHashAndMaskTop8Bits(const T* data, unsigned characterCount)
+    {
+        return finalizeAndMaskTop8Bits(computeHashImpl<T, Converter>(data, characterCount));
+    }
+
+    template<typename T>
+    static constexpr unsigned computeHashAndMaskTop8Bits(const T* data, unsigned characterCount)
+    {
+        return finalizeAndMaskTop8Bits(computeHashImpl<T>(data, characterCount));
+    }
+
+    template<typename T, unsigned charactersCount>
+    static constexpr unsigned computeLiteralHashAndMaskTop8Bits(const T (&characters)[charactersCount])
+    {
+        return computeHashAndMaskTop8Bits<T>(characters, charactersCount - 1);
+    }
+
+private:
+    static constexpr unsigned finalizeAndMaskTop8Bits(unsigned hash)
+    {
+        return avoidZero(hash & WYHasher::maskHash);
+    }
+
+    ALWAYS_INLINE static constexpr unsigned avoidZero(unsigned hash)
+    {
+        if (hash)
+            return hash;
+        return 0x80000000 >> flagCount;
+    }
+
+    ALWAYS_INLINE static constexpr uint64_t wyrot(uint64_t x)
+    {
+        return (x >> 32) | (x << 32);
+    }
+
+    ALWAYS_INLINE static constexpr void wymum(uint64_t* A, uint64_t* B)
+    {
+#if defined(__SIZEOF_INT128__)
+        __uint128_t r = *A;
+        r *= *B;
+        *A = static_cast<uint64_t>(r);
+        *B = static_cast<uint64_t>(r >> 64);
+#else
```

Comment:
Let's remove else side and always use `UInt128`. (Including `Int128.h`).
And because this is not C

```
UInt128 r = static_cast<UInt128>(*A) * *B;
*A = static_cast<uint64_t>(r);
*B = static_cast<uint64_t>(r >> 64);
```